Added support for NIOSSLCustomVerificationCallback for client connection

franck-clement-ug opened the pull request:
Hi,

This PR adds the ability for client connections to override the certificate verification logic implemented in nio-ssl.

The main goal is to allow client apps to perform their own certificate verification logic, such as SSL public key pinning.
Nio-ssl provide this feature as described here: [nio-ssl-issue-25](https://github.com/apple/swift-nio-ssl/issues/25) 
But I wasn't able to use it through grpc-swift because the `NIOSSLCustomVerificationCallback` is not accessible when initializing a secure `ClientConnection`.

I exposed a `NIOSSLCustomVerificationCallback` property in `ClientConnection.Configuration` struct and provided a `ClientConnection.Builder` func as well.

## Usage example for client public key pinning

```swift
let myPublicKey = Array(Data(base64Encoded: """
.... (Your base64 encoded public key)
""", options: .ignoreUnknownCharacters)!)

ClientConnection.secure(group: PlatformSupport.makeEventLoopGroup(loopCount: 1))
            .withTLSCustomVerificationCallback({ cert, eventLoopVerification in
                // Extract the leaf certificate's public key,
                // then compare it to the one you have.
                if let leaf = cert.first,
                   let publicKey = try? leaf.extractPublicKey() {

                    if let certSPKI = try? publicKey.toSPKIBytes(),
                       myPublicKey == certSPKI {
                        eventLoopVerification.succeed(.certificateVerified)
                    } else {
                        eventLoopVerification.fail(NIOSSLError.unableToValidateCertificate)
                    }
                } else {
                    eventLoopVerification.fail(NIOSSLError.noCertificateToValidate)
                }
            })
            .connect(host: host, port: port)
```

This PR + the example above should resolve #723 and resolve #1104

glbrntt commented on Sources/GRPC/ClientConnection.swift:
```diff
@@ -312,6 +312,9 @@ extension ClientConnection {
     }
 
+    /// A custom verification callback that allows completely overriding the certificate verification logic for this connection.
+    public var customVerificationCallback: NIOSSLCustomVerificationCallback?
```
Since this relates to TLS it should be on the `ClientConnection.Configuration.TLS` struct rather than the general configuration one.

glbrntt commented on Sources/GRPC/ClientConnection.swift:
```diff
@@ -426,16 +431,31 @@ extension Channel {
-      } catch {
-        return self.eventLoop.makeFailedFuture(error)
-      }
+        if let customVerificationCallback = customVerificationCallback {
```
There's a lot of duplicated code here: let's move the if-else inside the do-catch so that we don't duplicate the error handling or adding of handlers.

glbrntt commented on Sources/GRPC/GRPCChannel/GRPCChannelBuilder.swift:
```diff
@@ -273,6 +273,16 @@ extension ClientConnection.Builder {
   }
 }
 
+extension ClientConnection.Builder {
```
This should be an extension on `ClientConnection.Builder.Secure`: it doesn't apply to non-secure connections.

glbrntt commented on Sources/GRPC/GRPCChannel/GRPCChannelBuilder.swift:
```diff
@@ -273,6 +273,16 @@ extension ClientConnection.Builder {
+
+  /// A custom verification callback that allows completely overriding the certificate verification logic.
+  @discardableResult
+    public func withCustomVerification(callback: @escaping NIOSSLCustomVerificationCallback) -> Self {
```
This naming isn't clear at the call site:

```swift
builder.withCustomVerification { 
  // ...
}
```

I suggest: 

```swift
public func withTLSCustomVerificationCallback(_ callback: @escaping NIOSSLCustomVerificationCallback) -> Self {
  ...
}
```

glbrntt reviewed (changes requested):
Thanks for the PR, I've left some initial comments inline. You'll need to reformat the code to be consistent with the rest of the project (running `./scripts/format.sh` should do it). We'll also need a test for this; you should be able to use the `ClientTLSFailureTests` as a starting point.

franck-clement-ug commented:
@glbrntt Thank you for the detailed comments, they all make a lot of sense, and I fixed them. 
I also added a test when the validation logic is forced to fail within the `customVerificationCallback`.

glbrntt commented on Sources/GRPC/GRPCChannel/GRPCChannelBuilder.swift:
```diff
@@ -235,6 +235,16 @@ extension ClientConnection.Builder.Secure {
+
+  /// A custom verification callback that allows completely overriding the certificate verification logic.
+  @discardableResult
+  public func withCustomVerificationCallback(
```
This should be `withTLSCustomVerificationCallback`: without `TLS` it's unclear what is being verified.

glbrntt commented on Sources/GRPC/GRPCChannel/GRPCChannelBuilder.swift:
```diff
@@ -235,6 +235,16 @@ extension ClientConnection.Builder.Secure {
+  @discardableResult
+  public func withCustomVerificationCallback(
+    _ callback: @escaping NIOSSLCustomVerificationCallback
+  )
+    -> Self {
```
```suggestion
  ) -> Self {
```

glbrntt reviewed (changes requested):
Thanks @franck-clement-ug this looks good, just one naming nit to fix.

glbrntt reviewed (approved):
This looks good to me -- assuming CI passes -- thanks @franck-clement-ug!
